fcomm/src/bin/fcomm.rs

```diff
@@ -490,11 +496,87 @@ fn main() -> Result<(), Error> {
         .filter_level(cli.verbose.log_level_filter())
         .init();
 
+    // cache lives as long as the process
+    let cache = Box::leak(Box::new(cache::PublicParams {
+        file_store: Path::new(&cache::dir("lurk")).join("pubparams.json"),
+        data: HashMap::<usize, PublicParams>::new(),
+    }));
+
+    cache.read().ok();
+
     match &cli.command {
         Command::Commit(c) => c.commit(cli.limit),
-        Command::Open(o) => o.open(o.chain, cli.limit, cli.eval_input, o.quote_input),
+        Command::Open(o) => o.open(o.chain, cli.limit, cli.eval_input, o.quote_input, cache),
         Command::Eval(e) => e.eval(cli.limit),
-        Command::Prove(p) => p.prove(cli.limit),
-        Command::Verify(v) => v.verify(cli.error),
+        Command::Prove(p) => p.prove(cli.limit, cache),
+        Command::Verify(v) => v.verify(cli.error, cache),
+    }
+}
+
+mod cache {
+    use fcomm::error::Error;
+    use lurk::proof::nova;
+    use std::boxed::Box;
+    use std::collections::HashMap;
+    use std::env;
+    use std::hash::Hash;
+    use std::path::PathBuf;
+
+    pub struct Cache<
+        K: Eq + Hash + serde::Serialize + serde::de::DeserializeOwned,
+        V: serde::Serialize + serde::de::DeserializeOwned,
+    > {
+        pub data: HashMap<K, V>,
+        pub file_store: PathBuf,
+    }
+
+    pub type PublicParams<'a> = Cache<usize, nova::PublicParams<'a>>;
+
+    impl<
+            'a,
+            K: Eq + Hash + serde::Serialize + serde::de::DeserializeOwned,
+            V: serde::Serialize + serde::de::DeserializeOwned,
+        > Cache<K, V>
+    {
+        pub fn read(&mut self) -> Result<(), Error> {
+            let serialization = std::fs::read(self.file_store.clone())?;
+            self.data = serde_json::from_slice(serialization.as_slice())?;
+            Ok(())
+        }
+
+        pub fn write(&mut self) -> Result<(), Error> {
+            std::fs::create_dir_all(self.file_store.as_path().parent().unwrap())?;
+            let db = serde_json::to_string(&self.data)?;
+            std::fs::write(self.file_store.clone(), db)?;
```

Comment:
Instead of serializing the whole `HashMap` and reading/writing the entire cache, it would be better to serialize/deserialize each entry. Then only actually used cache values need to be loaded. This might not matter if we have only one thing cached (initially?), but it does in general.

In any case, I think you can/should use `FileMap` for this, since this is exactly the kind of thing it's for.

Reply:
Can you point me to what `FileMap` you mean? When I saw your comment last week, I assumed it was in the stdlib, but it's not. And the handful of crates that have a struct by this name, don't seem to be doing what we're after here.

Reply:
https://github.com/lurk-lang/lurk-rs/blob/master/fcomm/src/file_map.rs

As used throughout, for example: https://github.com/lurk-lang/lurk-rs/blob/master/fcomm/src/lib.rs#L65-L67 then https://github.com/lurk-lang/lurk-rs/blob/master/fcomm/src/lib.rs#L648-L651